### PR TITLE
Add FontEditor tests, fix bugs and debug asserts

### LINKED ISSUE #761: FontEditor.EditValue for an invalid provider can throw InvalidCastException
E.g.
```cs
using Moq;

public static IEnumerable<object[]> EditValue_InvalidProvider_TestData()
{
    yield return new object[] { null };

    var nullServiceProviderMock = new Mock<IServiceProvider>(MockBehavior.Strict);
    nullServiceProviderMock
        .Setup(p => p.GetService(typeof(IWindowsFormsEditorService)))
        .Returns(null);
    yield return new object[] { nullServiceProviderMock.Object };

    var invalidServiceProviderMock = new Mock<IServiceProvider>(MockBehavior.Strict);
    invalidServiceProviderMock
        .Setup(p => p.GetService(typeof(IWindowsFormsEditorService)))
        .Returns(new object());
    yield return new object[] { invalidServiceProviderMock.Object };
}

[Theory]
[MemberData(nameof(EditValue_InvalidProvider_TestData))]
public void EditValue_InvalidProvider_ReturnsValue(IServiceProvider provider)
{
    var editor = new FontEditor();
    var value = new object();
    Assert.Same(value, editor.EditValue(null, provider, value));
}
```

If the service is not `IWindowsFormsEditorService` then an `InvalidCastException` is thrown.

Proposed fix is to use an `as` cast

### PR DESCRIPTION
**Pull request by hughbe**
Fixes #761

### CODE REVIEW

File: src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs [hunk 1]
```diff
@@ -11,53 +11,42 @@
...
+        /// A service provider is provided so that any required editing services can be obtained.
         /// </summary>
         public override object EditValue(ITypeDescriptorContext context, IServiceProvider provider, object value)
         {
-            this.value = value;
-
-            Debug.Assert(provider != null, "No service provider; we cannot edit the value");
             if (provider != null)
             {
-                IWindowsFormsEditorService edSvc = (IWindowsFormsEditorService)provider.GetService(typeof(IWindowsFormsEditorService));
```
Comment (weltkante) [inline src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: Isn't that a common idiom that a service for an interface implements that interface? I've always considered it a bug in the caller if he provided a bogus service provider, not being worth programming defensively against.

You'll definitively have a lot of places to "fix" this, I'd assume most GetService calls are a strict cast and not an as-cast with fallback code

Comment (zsd4yr) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: I agree with @weltkante, this is too large a change and the paradigm is too spread out. Please undo this line change

File: src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs [hunk 2]
```diff
@@ -11,53 +11,42 @@
...
                     }
 
                     IntPtr hwndFocus = UnsafeNativeMethods.GetFocus();
                     try
                     {
-                        if (fontDialog.ShowDialog() == DialogResult.OK)
+                        if (_fontDialog.ShowDialog() == DialogResult.OK)
                         {
-                            this.value = fontDialog.Font;
+                            return _fontDialog.Font;
```
Comment (zsd4yr) [inline src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: Won't this prevent the finally from occurring?

Comment (hughbe) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: https://stackoverflow.com/questions/10255774/if-i-return-out-of-a-try-finally-block-in-c-sharp-does-the-code-in-the-finally-a?lq=1

it shall run!

File: src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs (hunk 1 above)
Comment (hughbe) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: I don't agree here so I'll try to explain what my logic is.
Across winforms we have about a 50-50 split between using `(X)provider.GetService(typeof(X))` and using `provider.GetService(typeof(x)) as X`. I think the use of `as/is` has numerous benefits
- Clears up confusion as to what happens when the service is `null` -  will this throw `InvalidCastException` or `NullReferenceException`. Actually, it's neither - it returns `null` - but the `as/is` pattern avoids this confusion.
- Most importantly, we can use the C# construct `if (variable is Type variableAsType)` which lets us made code more readable, express intent better etc.

It's essentially my view is not that we want to throw `InvalidCastException` - if this were the case we'd use the force cast rather than the `as` cast across winforms, which we don't. The intent to get an object of that type. In each case we check for null so we understand that the provider can return an invalid service. In some cases we've accounted for where the provider returns a non-interface service, in others we haven't. It's my opinion that we should standardise on the latter.

Comment (zsd4yr) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: @hughbe my suggestion is to 
1. remove this change from this PR
1. create an issue discussing the discrepancy in the code base, your suggested change, and its benefits
1. await a decision from them team after broader community discussion

and then if the decision is to change these more broadly (I suspect we will want to align them all one way or the other), then that change should come in its own PR.

File: src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs (hunk 2 above)
Comment (zsd4yr) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: I see https://stackoverflow.com/questions/421797/what-really-happens-in-a-try-return-x-finally-x-null-statement

I am trying to weigh the pros and cons here. On one hand:
* this return adds more clarity be removing the need to use an additional variable

However, on the other hand:
* this return adds confusion because the underlying IL and assembly will create a local variable, so this code may not do what a programmer expects it to do (for example, it was not straight-forward to me)

Thoughts?

@jaredpar do you have any opinions about c# code cleanliness versus eventual IL/assembly code clarity here?

Comment (hughbe) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: I don't see this as a problem - essentially a using statement is the same/similar to a try/finally block and we don't have problems returning from inside usings! TBH I don't know much about compiler internals, so am very read to be proved wrong!

File: src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs (hunk 1 above)
Comment (hughbe) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: Just noting that previously we had a `Debug` assert here suggesting that we would never expect an invalid service to be returned from `provider.GetService`. In essence I've uncovered some edge cases that are highly unlikely to occur in practice, where the designer makes things valid. However, I still think this sort of test coverage is important :)))

```cs
Debug.Assert(provider != null, "No service provider; we cannot edit the value");
if (provider != null)
{
    IWindowsFormsEditorService edSvc = (IWindowsFormsEditorService)provider.GetService(typeof(IWindowsFormsEditorService));

    Debug.Assert(edSvc != null, "No editor service; we cannot edit the value");
    if (edSvc != null)
    {
        ...
    }
}
```

File: src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs (hunk 2 above)
Comment (JuditRose) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: I agree with @hughbe in this question, by definition ```finally``` block will run after ```try```.
"The transfer of control can occur as a result of normal execution, of execution of a ```break```, ```continue```, ```goto```, or ```return``` statement, or of propagation of an exception out of the try statement." (https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/try-finally)

Comment (zsd4yr) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: I understand that the finally will always run; I am arguing that it is not _clear_ that _any_ code will run post return within the same scope.

@hughbe when you return inside a using statement, the method ends. Dispose will be called, but no code internal to that function's scope will be run. 

I don't like it.

Comment (hughbe) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: IMO I don't see putting returns inside try/finally blocks as a problem - it just is a common code pattern I've seen across various dotnet repos

File: src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs (hunk 1 above)
Comment (hughbe) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: Now reverted the change, as per #817 decision

File: src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs (hunk 2 above)
Comment (RussKie) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: Agree, any developer worth his/her salt is expected to understand this. Especially working on a framework repo.

Comment (zsd4yr) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: @RussKie to be clear, I did not initially understand this. In fact, I feel strongly that this may violate some calling conventions. Strictly nothing in a given scope (the method's scope) should run after a return statement. Likely this try-finally case and the case with using are a concession around the competing feature work of those areas against the calling convention.

Since this finally is relatively short and does not appear affect the types in the try or catch, I think the early return is fine.

Comment (weltkante) [reply src/System.Windows.Forms.Design.Editors/src/System/Drawing/Design/FontEditor.cs]: @zsd4yr finally works like this in almost all programming languages, I think it is reasonable to expect someone working on a core Framework library to understand the concept. Also it has nothing to do with calling conventions, the finally is resolved entirely within the method body by the compiler/runtime and not visible to the caller.
